docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+- `checksum` is a checksum of the expected result of a request to the `url`.
+  It is not required, but is recommended.
+  If provided, it MUST be a hex encoded checksum string.
+  Tools utilizing these `UpgradeInstructions` MUST fail if a `checksum` is provided but is different from the checksum of the result returned by the `url`.
+- `checksum_algo` is a string identify the algorithm used to generate the `checksum`.
+  Recommended algorithms: `sha256`, `sha512`.
+  Algorithms also supported (but not recommended): `sha1`, `md5`.
+  If a `checksum` is provided, a `checksum_algo` MUST also be provided.
+
+A `url` is not required to contain a `checksum` query parameter.
+If the `url` does contain a `checksum` query parameter, the `checksum` and `checksum_algo` fields MUST also be populated, and their values MUST match the value of the query parameter.
+For example, if the `url` is `"https://example.com?checksum=md5:d41d8cd98f00b204e9800998ecf8427e"`, then the `checksum` field must be `"d41d8cd98f00b204e9800998ecf8427e"` and the `checksum_algo` field must be `"md5"`.
+
+### Upgrade Module Updates
+
+If an upgrade `Plan` does not use the new `UpgradeInstructions` field, existing functionality will be maintained.
+
+We will update the creation of the `upgrade-info.json` file to include the `UpgradeInstructions`.
+
+We will update the optional validation available via CLI to account for the new `Plan` structure.
+We will add the following validation:
+1.  If `UpgradeInstructions` are provided:
+    1.  There MUST be at least one entry in `artifacts`.
+    1.  All of the `artifacts` MUST have a unique `platform`.
+    1.  For each `Artifact`, if the `url` contains a `checksum` query parameter:
+        1. The `checksum` query parameter value MUST be in the format of `{checksum_algo}:{checksum}`.
+        1. The `{checksum}` from the query parameter MUST equal the `checksum` provided in the `Artifact`.
+        1. The `{checksum_algo}` from the query parameter MUST equal the `checksum_algo` provided in the `Artifact`.
+1.  The following validation is currently done using the `info` field. We will apply similar validation to the `UpgradeInstructions`.
+    For each `Artifact`:
+    1.  The `platform` MUST have the format `{OS}/{CPU}` or be `"any"`.
+    1.  The `url` field MUST NOT be empty.
+    1.  The `url` field MUST be a proper URL.
+    1.  A `checksum` MUST be provided either in the `checksum` field or as a query parameter in the `url`.
+    1.  If the `checksum` field has a value and the `url` also has a `checksum` query parameter, the two values MUST be equal.
+    1.  The `url` MUST return either a file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
+    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.
+
```

Comment:
```suggestion
1. `Plan.Info` will be deprecated - tools and CLI should return a warning if `Plan.Info` is provided but `UpgradeInstructions` is not.

```

Reply:
I've held off on making this change because I'd still like to have a way to provide a URL to JSON of upgrade instructions.

If we want to fully deprecate the `info` field, then we'd need to add a new field specifically for such a URL.

We could also use a `one of` for the `instructions`, to make it clear that you either define the instructions in the message or provide a url to them. But I'm not a fan of the `one of` stuff.

Reply:
This other conversation was similar to this one. I "resolved" that one and linked to this one.

https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r782472640

Reply:
Sorry to chain this further, but this discussion is also going on here https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r781438288. I will close this one and continue it there.